### DISCUSSION TITLE
brew services list results in Error: undefined method `length' for nil:NilClass

### DISCUSSION THREAD

**Discussion by Konstrukteur** [Homebrew/discussions#3112] [Linux]
I am using Homebrew on Linux 20.04. After apt-get update and upgrade and brew update and upgrade I am unable to check status of installed programs. I am also unable to get mysql, installed through homebrew, running,

    % brew services list      
    Error: undefined method `length' for nil:NilClass
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-services/lib/service/commands/list.rb:50:in `block in print_table'
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-services/lib/service/commands/list.rb:50:in `map'
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-services/lib/service/commands/list.rb:50:in `print_table'
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-services/lib/service/commands/list.rb:22:in `run'
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-services/cmd/services.rb:93:in `services'
    /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:110:in `<main>'

After updating I was unable to do a brew upgrade and I discovered that the permissions on the directories inside /home/.linuxbrew were set between two users. I changed them to my main user account and was able to run brew upgrade. But I am still unable to get mysql running or get the status of the installed services.

Is there any bug on the new version of homebrew/linuxbrew? Or how can I get it running?

Thank you in advance.

**Comment by SMillerDev** [comment]
what does `brew services info --all` print?

**Comment by Konstrukteur** [comment]
Dear Sean Molenaar, thank you for your reply. I lost a day trying to repair the system. The whole setting had stopped working. At the end I resorted to restoring the VPS from backup and doing a `brew update` and `brew upgrade` after having unified the permissions to one user account (it was set between two deploy user accounts). For that reason I "fortunately" cannot replicate the issue and run the command. To your info: Brew had become totally unusable. I was unable to start any services and could not run `brew services list`.

Now the server is running again, but I still haven't run an `apt-get update`.

I do not know, what went wrong. Maybe permissions... I still stumble into OpenSSL problems here and there.

If I may ask. For compilation of Ruby Gems etc. do I have to declare PATH, LDFLAGS, etc in my shell? It is what I usually do, but do not understand completely how it works. I would like to get a better grasp on it, since I suspect it could be one of the causes.

    ### Customize path for openssl
    export PATH="/home/linuxbrew/.linuxbrew/opt/openssl@3/bin:$PATH"
    export LDFLAGS="-L/home/linuxbrew/.linuxbrew/opt/openssl@3/lib"
    export CPPFLAGS="-I/home/linuxbrew/.linuxbrew/opt/openssl@3/include"
    LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/linuxbrew/.linuxbrew/Cellar/openssl@3/3.0.2/lib/libssl.so.3

**Reply by SMillerDev** [reply]
> If I may ask. For compilation of Ruby Gems etc. do I have to declare PATH, LDFLAGS, etc in my shell? It is what I usually do, but do not understand completely how it works. I would like to get a better grasp on it, since I suspect it could be one of the causes.

That's not something that Homebrew generally manages so I'm not sure.

**Comment by Konstrukteur** [comment]
I try to rephrase: When installing formulas through Homebrew, sometimes dependencies are installed. I use Homebrew for MySQL, Ruby, Nginx, etc. When a formula has a dependency such as for example OpenSSL, I always thought it would need the info about this dependency and added the relative Paths and Flags to my shell.

I asked because in one default gemset I still get, after the restore, an SSL related error and when troubleshooting my brew install, before the restore, I had brew complain about its CA-certificates, etc ...

**Reply by SMillerDev** [reply]
> When a formula has a dependency such as for example OpenSSL, I always thought it would need the info about this dependency and added the relative Paths and Flags to my shell.

No, Homebrew manages it's own dependencies. No need to add anything.

> I asked because in one default gemset I still get, after the restore, an SSL related error

Is it a formula or a gem? If it's the former that's a brew issue, if it's the latter I don't think I can help much.

**Comment by Konstrukteur** [comment]
Ok, the formula does solve it's dependencies. Which means, that when I get an error related to SSL, when trying to install an application through brew, it should be a problem related to the system setting? And not the fact that I have for example the wrong 
LDFLAGS of a brew installed OpenSSL version in my shell?

What I do not understand is, that some formulas apparently depend on openssl@1.1 and others on openssl@3. (At least I have both without installing them explicitly.) Does it mean that the dependency is related to compiling the formula and not to the functioning of the compiled application?

I will try to get to the cause of the Ruby issue over at RVM.

Thank you very much for your insight!

**Reply by SMillerDev** [reply]
> What I do not understand is, that some formulas apparently depend on openssl@1.1 and others on openssl@3. (At least I have both without installing them explicitly.) Does it mean that the dependency is related to compiling the formula and not to the functioning of the compiled application?

They are related to both. OpenSSL is usually required at compile time and at runtime, but as an example: for Go programs Go is usually not required at runtime.

**Reply by Konstrukteur** [reply]
Thanks for your feedback. The fact that I ended up having multiple openssl versions through brew and maybe even through apt-get causes me headache, when I have to troubleshoot. I think now I have a little bit more insight into the piping of my systems. Thank you for everything!
